plugins/repository-gcs/src/main/java/io/crate/gcs/GCSBlobStore.java

```diff
@@ -54,11 +104,283 @@ public BlobContainer blobContainer(BlobPath path) {
+        }
+        if (length < 0) {
+            throw new IllegalArgumentException("length must be non-negative");
+        }
+        if (length == 0) {
+            return new ByteArrayInputStream(new byte[0]);
+        } else {
+            return new GCSRetryingInputStream(
+                client(),
+                BlobId.of(bucketName, blobName),
+                position,
+                Math.addExact(position, length - 1)
+            );
+        }
+    }
+
+    /**
+     * Writes a blob in the specific bucket
+     *
+     * @param inputStream         content of the blob to be written
+     * @param blobSize            expected size of the blob to be written
+     * @param failIfAlreadyExists whether to throw a FileAlreadyExistsException if the given blob already exists
+     */
+    void writeBlob(String blobName, InputStream inputStream, long blobSize, boolean failIfAlreadyExists) throws IOException {
+        final BlobInfo blobInfo = BlobInfo.newBuilder(bucketName, blobName).build();
+        if (blobSize > getLargeBlobThresholdInBytes()) {
+            writeBlobResumable(blobInfo, inputStream, blobSize, failIfAlreadyExists);
+        } else {
+            writeBlobMultipart(blobInfo, inputStream, blobSize, failIfAlreadyExists);
+        }
+    }
+
+    long getLargeBlobThresholdInBytes() {
+        return LARGE_BLOB_THRESHOLD_BYTE_SIZE;
+    }
+
+    /**
+     * Uploads a blob using the "resumable upload" method (multiple requests, which
+     * can be independently retried in case of failure, see
+     * https://cloud.google.com/storage/docs/json_api/v1/how-tos/resumable-upload
+     *
+     * @param blobInfo            the info for the blob to be uploaded
+     * @param inputStream         the stream containing the blob data
+     * @param size                expected size of the blob to be written
+     * @param failIfAlreadyExists whether to throw a FileAlreadyExistsException if the given blob already exists
+     */
+    private void writeBlobResumable(BlobInfo blobInfo, InputStream inputStream, long size, boolean failIfAlreadyExists) throws IOException {
+        // We retry 410 GONE errors to cover the unlikely but possible scenario where a resumable upload session becomes broken and
+        // needs to be restarted from scratch. Given how unlikely a 410 error should be according to SLAs we retry only twice.
+        assert inputStream.markSupported();
+        inputStream.mark(Integer.MAX_VALUE);
+        final byte[] buffer = new byte[size < bufferSize ? Math.toIntExact(size) : bufferSize];
+        StorageException storageException = null;
+        final Storage.BlobWriteOption[] writeOptions = failIfAlreadyExists
+            ? new Storage.BlobWriteOption[]{Storage.BlobWriteOption.doesNotExist()}
+            : new Storage.BlobWriteOption[0];
+        for (int retry = 0; retry < 3; ++retry) {
+            try {
+                final WriteChannel writeChannel = client().writer(blobInfo, writeOptions);
```

Comment:
writer should be closed.
```suggestion
            try (WriteChannel writeChannel = client().writer(blobInfo, writeOptions))
```